File: packages/agroclima/geolocation/resources/android/AgroClimaGeolocationFunctions.kt
package com.agroclima.plugins.geolocation

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONObject

object AgroClimaGeolocationFunctions {
    private const val LOCATION_EVENT = "Native\\Mobile\\Events\\Geolocation\\LocationReceived"
    private const val PERMISSION_STATUS_EVENT =
        "Native\\Mobile\\Events\\Geolocation\\PermissionStatusReceived"
    private const val PERMISSION_REQUEST_EVENT =
        "Native\\Mobile\\Events\\Geolocation\\PermissionRequestResult"

    class GetCurrentPosition(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.runOnUiThread {
                GeolocationCoordinator.install(activity).getCurrentPosition(
                    PositionRequest(
                        id = parameters["id"] as? String,
                        event = parameters["event"] as? String ?: LOCATION_EVENT,
                        fineAccuracy = parameters["fineAccuracy"] as? Boolean ?: false,
                    ),
                )
            }

            return BridgeResponse.success(emptyMap<String, Any>())
        }
    }

    class CheckPermissions(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.runOnUiThread {
                val event = parameters["event"] as? String ?: PERMISSION_STATUS_EVENT
                val id = parameters["id"] as? String
                GeolocationCoordinator.install(activity).checkPermissions(event, id)
            }

            return BridgeResponse.success(emptyMap<String, Any>())
        }
    }

    class RequestPermissions(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.runOnUiThread {
                GeolocationCoordinator.install(activity).requestPermissions(
                    PermissionRequest(
                        id = parameters["id"] as? String,
                        event = parameters["event"] as? String ?: PERMISSION_REQUEST_EVENT,
                    ),
                )
            }

            return BridgeResponse.success(emptyMap<String, Any>())
        }
    }
}

data class PositionRequest(
    val id: String?,
    val event: String,
    val fineAccuracy: Boolean,
)

data class PermissionRequest(
    val id: String?,
    val event: String,
)

class GeolocationCoordinator : Fragment(), LocationListener {
    private val handler = Handler(Looper.getMainLooper())
    private val positionRequests = mutableListOf<PositionRequest>()
    private val permissionRequests = mutableListOf<PermissionRequest>()
    private var permissionRequestInFlight = false
    private var locationRequestInFlight = false
    private var activeProvider: String? = null

    private val permissionLauncher =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) {
            permissionRequestInFlight = false

            permissionRequests.toList().forEach { request ->
                dispatchPermissionState(request.event, request.id, requestResult = true)
            }
            permissionRequests.clear()

            if (hasAnyLocationPermission()) {
                startLocationRequest()
            } else {
                failPositions("El permiso de ubicación fue denegado.")
            }
        }

    private val timeout = Runnable {
        failPositions("No fue posible obtener la ubicación a tiempo.")
    }

    private val preferences by lazy {
        requireContext().getSharedPreferences("agroclima_geolocation", Context.MODE_PRIVATE)
    }

    private val locationManager by lazy {
        requireContext().getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }

    fun getCurrentPosition(request: PositionRequest) {
        if (!hasAnyLocationPermission()) {
            if (hasRequestedPermission()) {
                dispatchLocationError(request, "El permiso de ubicación está denegado.")
                return
            }

            positionRequests += request
            launchPermissionRequest()
            return
        }

        positionRequests += request
        startLocationRequest()
    }

    fun checkPermissions(event: String, id: String?) {
        dispatchPermissionState(event, id, requestResult = false)
    }

    fun requestPermissions(request: PermissionRequest) {
        if (hasAnyLocationPermission()) {
            dispatchPermissionState(request.event, request.id, requestResult = true)
            return
        }

        if (isPermanentlyDenied()) {
            dispatchPermissionState(request.event, request.id, requestResult = true)
            return
        }

        permissionRequests += request
        launchPermissionRequest()
    }

    private fun launchPermissionRequest() {
        if (permissionRequestInFlight) {
            return
        }

        permissionRequestInFlight = true
        preferences.edit().putBoolean(KEY_PERMISSION_REQUESTED, true).apply()
        permissionLauncher.launch(
            arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION,
            ),
        )
    }

    private fun startLocationRequest() {
        if (positionRequests.isEmpty() || locationRequestInFlight) {
            return
        }

        val wantsFineAccuracy = positionRequests.any(PositionRequest::fineAccuracy)
        val provider = chooseProvider(wantsFineAccuracy)

        if (provider == null) {
            failPositions("No hay un proveedor de ubicación compatible con los permisos concedidos.")
            return
        }

        try {
            activeProvider = provider
            locationRequestInFlight = true
            locationManager.requestLocationUpdates(
                provider,
                0L,
                0f,
                this,
                Looper.getMainLooper(),
            )
            handler.postDelayed(timeout, LOCATION_TIMEOUT_MS)
        } catch (_: SecurityException) {
            failPositions("El permiso de ubicación no está disponible.")
        } catch (_: IllegalArgumentException) {
            failPositions("No hay un proveedor de ubicación disponible.")
        }
    }

    private fun chooseProvider(fineAccuracy: Boolean): String? {
        val hasFinePermission = hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)

        if (fineAccuracy && hasFinePermission && isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            return LocationManager.GPS_PROVIDER
        }

        if (isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
            return LocationManager.NETWORK_PROVIDER
        }

        if (hasFinePermission && isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            return LocationManager.GPS_PROVIDER
        }

        return null
    }

    private fun isProviderEnabled(provider: String): Boolean =
        try {
            locationManager.isProviderEnabled(provider)
        } catch (_: Exception) {
            false
        }

    override fun onLocationChanged(location: Location) {
        handler.removeCallbacks(timeout)
        stopLocationUpdates()
        locationRequestInFlight = false
        activeProvider = null

        positionRequests.toList().forEach { request ->
            val payload = JSONObject().apply {
                put("success", true)
                put("latitude", location.latitude)
                put("longitude", location.longitude)
                put("accuracy", location.accuracy.toDouble())
                put("timestamp", location.time)
                put("provider", location.provider ?: "android")
                request.id?.let { put("id", it) }
            }
            dispatch(request.event, payload)
        }
        positionRequests.clear()
    }

    override fun onProviderDisabled(provider: String) {
        if (provider == activeProvider) {
            failPositions("El proveedor de ubicación fue desactivado.")
        }
    }

    @Deprecated("Required by LocationListener on older Android versions")
    override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) = Unit

    private fun failPositions(message: String) {
        handler.removeCallbacks(timeout)
        if (locationRequestInFlight) {
            stopLocationUpdates()
        }
        locationRequestInFlight = false
        activeProvider = null

        positionRequests.toList().forEach { dispatchLocationError(it, message) }
        positionRequests.clear()
    }

    private fun dispatchLocationError(request: PositionRequest, message: String) {
        val payload = JSONObject().apply {
            put("success", false)
            put("error", message)
            request.id?.let { put("id", it) }
        }
        dispatch(request.event, payload)
    }

    private fun dispatchPermissionState(event: String, id: String?, requestResult: Boolean) {
        val fineGranted = hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)
        val coarseGranted = hasPermission(Manifest.permission.ACCESS_COARSE_LOCATION)
        val fallback = permissionFallback(requestResult)

        val payload = JSONObject().apply {
            put("location", if (fineGranted || coarseGranted) GRANTED else fallback)
            put("coarseLocation", if (coarseGranted) GRANTED else fallback)
            put("fineLocation", if (fineGranted) GRANTED else fallback)
            id?.let { put("id", it) }
        }
        dispatch(event, payload)
    }

    private fun permissionFallback(requestResult: Boolean): String {
        if (!hasRequestedPermission()) {
            return NOT_DETERMINED
        }

        if (requestResult && isPermanentlyDenied()) {
            return PERMANENTLY_DENIED
        }

        return DENIED
    }

    private fun isPermanentlyDenied(): Boolean =
        hasRequestedPermission() &&
            !hasAnyLocationPermission() &&
            !shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_FINE_LOCATION) &&
            !shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION)

    private fun hasRequestedPermission(): Boolean =
        preferences.getBoolean(KEY_PERMISSION_REQUESTED, false)

    private fun hasAnyLocationPermission(): Boolean =
        hasPermission(Manifest.permission.ACCESS_FINE_LOCATION) ||
            hasPermission(Manifest.permission.ACCESS_COARSE_LOCATION)

    private fun hasPermission(permission: String): Boolean =
        ContextCompat.checkSelfPermission(requireContext(), permission) ==
            PackageManager.PERMISSION_GRANTED

    private fun dispatch(event: String, payload: JSONObject) {
        NativeActionCoordinator.dispatchEvent(requireActivity(), event, payload.toString())
    }

    private fun stopLocationUpdates() {
        try {
            locationManager.removeUpdates(this)
        } catch (_: SecurityException) {
            // Permission can be revoked while a request is in flight.
        }
    }

    override fun onDestroy() {
        handler.removeCallbacks(timeout)
        if (locationRequestInFlight) {
            stopLocationUpdates()
        }
        super.onDestroy()
    }

    companion object {
        private const val FRAGMENT_TAG = "AgroClimaGeolocationCoordinator"
        private const val LOCATION_TIMEOUT_MS = 15_000L
        private const val KEY_PERMISSION_REQUESTED = "permission_requested"
        private const val GRANTED = "granted"
        private const val DENIED = "denied"
        private const val NOT_DETERMINED = "not_determined"
        private const val PERMANENTLY_DENIED = "permanently_denied"

        fun install(activity: FragmentActivity): GeolocationCoordinator =
            activity.supportFragmentManager.findFragmentByTag(FRAGMENT_TAG) as? GeolocationCoordinator
                ?: GeolocationCoordinator().also { fragment ->
                    activity.supportFragmentManager.beginTransaction()
                        .add(fragment, FRAGMENT_TAG)
                        .commitNowAllowingStateLoss()
                }
    }
}

File: packages/agroclima/geolocation/resources/ios/AgroClimaGeolocationFunctions.swift
import CoreLocation
import Foundation

enum AgroClimaGeolocationFunctions {
    private static let locationEvent = "Native\\Mobile\\Events\\Geolocation\\LocationReceived"
    private static let permissionStatusEvent =
        "Native\\Mobile\\Events\\Geolocation\\PermissionStatusReceived"
    private static let permissionRequestEvent =
        "Native\\Mobile\\Events\\Geolocation\\PermissionRequestResult"

    final class GetCurrentPosition: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let request = PositionRequest(
                id: parameters["id"] as? String,
                event: parameters["event"] as? String ?? locationEvent,
                fineAccuracy: parameters["fineAccuracy"] as? Bool ?? false
            )

            DispatchQueue.main.async {
                GeolocationCoordinator.shared.getCurrentPosition(request)
            }

            return BridgeResponse.success(data: [:])
        }
    }

    final class CheckPermissions: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let event = parameters["event"] as? String ?? permissionStatusEvent
            let id = parameters["id"] as? String

            DispatchQueue.main.async {
                GeolocationCoordinator.shared.checkPermissions(event: event, id: id)
            }

            return BridgeResponse.success(data: [:])
        }
    }

    final class RequestPermissions: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let request = PermissionRequest(
                id: parameters["id"] as? String,
                event: parameters["event"] as? String ?? permissionRequestEvent
            )

            DispatchQueue.main.async {
                GeolocationCoordinator.shared.requestPermissions(request)
            }

            return BridgeResponse.success(data: [:])
        }
    }
}

private struct PositionRequest {
    let id: String?
    let event: String
    let fineAccuracy: Bool
}

private struct PermissionRequest {
    let id: String?
    let event: String
}

private final class GeolocationCoordinator: NSObject, CLLocationManagerDelegate {
    static let shared = GeolocationCoordinator()

    private let manager = CLLocationManager()
    private var positionRequests: [PositionRequest] = []
    private var permissionRequests: [PermissionRequest] = []
    private var timeout: DispatchWorkItem?
    private var locationRequestInFlight = false

    private override init() {
        super.init()
        manager.delegate = self
    }

    func getCurrentPosition(_ request: PositionRequest) {
        switch manager.authorizationStatus {
        case .notDetermined:
            positionRequests.append(request)
            manager.requestWhenInUseAuthorization()
        case .authorizedAlways, .authorizedWhenInUse:
            positionRequests.append(request)
            startLocationRequest()
        case .denied:
            dispatchLocationError(request, message: "El permiso de ubicación está denegado.")
        case .restricted:
            dispatchLocationError(request, message: "La ubicación está restringida en este dispositivo.")
        @unknown default:
            dispatchLocationError(request, message: "No fue posible determinar el permiso de ubicación.")
        }
    }

    func checkPermissions(event: String, id: String?) {
        dispatchPermissionState(event: event, id: id, requestResult: false)
    }

    func requestPermissions(_ request: PermissionRequest) {
        if manager.authorizationStatus == .notDetermined {
            permissionRequests.append(request)
            manager.requestWhenInUseAuthorization()
            return
        }

        dispatchPermissionState(event: request.event, id: request.id, requestResult: true)
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        guard manager.authorizationStatus != .notDetermined else {
            return
        }

        permissionRequests.forEach { request in
            dispatchPermissionState(event: request.event, id: request.id, requestResult: true)
        }
        permissionRequests.removeAll()

        switch manager.authorizationStatus {
        case .authorizedAlways, .authorizedWhenInUse:
            startLocationRequest()
        case .denied:
            failPositions("El permiso de ubicación fue denegado.")
        case .restricted:
            failPositions("La ubicación está restringida en este dispositivo.")
        case .notDetermined:
            break
        @unknown default:
            failPositions("No fue posible determinar el permiso de ubicación.")
        }
    }

    private func startLocationRequest() {
        guard !positionRequests.isEmpty, !locationRequestInFlight else {
            return
        }

        manager.desiredAccuracy = positionRequests.contains(where: \.fineAccuracy)
            ? kCLLocationAccuracyBest
            : kCLLocationAccuracyKilometer
        locationRequestInFlight = true
        manager.requestLocation()

        let timeout = DispatchWorkItem { [weak self] in
            self?.failPositions("No fue posible obtener la ubicación a tiempo.")
        }
        self.timeout = timeout
        DispatchQueue.main.asyncAfter(deadline: .now() + 15, execute: timeout)
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else {
            failPositions("El dispositivo no devolvió una ubicación válida.")
            return
        }

        timeout?.cancel()
        timeout = nil
        locationRequestInFlight = false

        let requests = positionRequests
        positionRequests.removeAll()

        requests.forEach { request in
            var payload: [String: Any] = [
                "success": true,
                "latitude": location.coordinate.latitude,
                "longitude": location.coordinate.longitude,
                "accuracy": location.horizontalAccuracy,
                "timestamp": Int(location.timestamp.timeIntervalSince1970 * 1_000),
                "provider": "corelocation",
            ]
            if let id = request.id {
                payload["id"] = id
            }
            dispatch(event: request.event, payload: payload)
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        let locationError = error as? CLError
        let message = locationError?.code == .denied
            ? "El permiso o los servicios de ubicación no están disponibles."
            : "No fue posible obtener la ubicación actual."
        failPositions(message)
    }

    private func failPositions(_ message: String) {
        timeout?.cancel()
        timeout = nil
        locationRequestInFlight = false

        let requests = positionRequests
        positionRequests.removeAll()
        requests.forEach { dispatchLocationError($0, message: message) }
    }

    private func dispatchLocationError(_ request: PositionRequest, message: String) {
        var payload: [String: Any] = [
            "success": false,
            "error": message,
        ]
        if let id = request.id {
            payload["id"] = id
        }
        dispatch(event: request.event, payload: payload)
    }

    private func dispatchPermissionState(event: String, id: String?, requestResult: Bool) {
        let status = manager.authorizationStatus
        let authorized = status == .authorizedAlways || status == .authorizedWhenInUse
        let fineAuthorized = authorized && manager.accuracyAuthorization == .fullAccuracy
        let fallback = permissionFallback(status: status, requestResult: requestResult)

        var payload: [String: Any] = [
            "location": authorized ? "granted" : fallback,
            "coarseLocation": authorized ? "granted" : fallback,
            "fineLocation": fineAuthorized ? "granted" : fallback,
        ]
        if let id {
            payload["id"] = id
        }
        dispatch(event: event, payload: payload)
    }

    private func permissionFallback(
        status: CLAuthorizationStatus,
        requestResult: Bool
    ) -> String {
        switch status {
        case .notDetermined:
            return "not_determined"
        case .denied:
            return requestResult ? "permanently_denied" : "denied"
        case .restricted:
            return "denied"
        case .authorizedAlways, .authorizedWhenInUse:
            return "denied"
        @unknown default:
            return "denied"
        }
    }

    private func dispatch(event: String, payload: [String: Any]) {
        dispatchPrecondition(condition: .onQueue(.main))
        LaravelBridge.shared.send?(event, payload)
    }
}
